packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx

```diff
@@ -39,12 +65,39 @@ export type GridToolbarQuickFilterProps = TextFieldProps & {
 };
 
 function GridToolbarQuickFilter(props: GridToolbarQuickFilterProps) {
-  const { quickFilterParser = defaultSearchValueParser, debounceMs = 500, ...other } = props;
+  const {
+    quickFilterParser = defaultSearchValueParser,
+    quickFilterFormatter = defaultSearchValueFormatter,
+    debounceMs = 500,
+    ...other
+  } = props;
 
   const apiRef = useGridApiContext();
   const rootProps = useGridRootProps();
+  const quickFilterValues = gridQuickFilterValuesSelector(apiRef);
 
-  const [searchValue, setSearchValue] = React.useState('');
+  const [searchValue, setSearchValue] = React.useState(
+    quickFilterFormatter(quickFilterValues ?? []),
+  );
+
+  const [prevQuickFilterValues, setPrevQuickFilterValues] = React.useState(quickFilterValues);
+
+  React.useEffect(() => {
+    if (!isDeepEqual(prevQuickFilterValues, quickFilterValues)) {
+      // The model of quick filter value has been updated
+      setPrevQuickFilterValues(quickFilterValues);
+      if (!isDeepEqual(quickFilterParser(searchValue), quickFilterValues)) {
+        // The input value and the new model are un-sync so we will update the input
+        setSearchValue(quickFilterFormatter(quickFilterValues ?? []));
```

Comment:
It's calling this effect twice. You can do two things to improve it:
- use a ref for `prevQuickFilterValues`, since it's not part of the state of the component
- instead of `setSearchValue` do `setSearchValue(prevSearchValue => { ... })` to remove `searchValue` from the dependencies.